include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp

```diff
@@ -0,0 +1,156 @@
+#ifndef LBANN_EXECUTION_ALGORITHMS_LTFB_MUTATION_STRATEGY_HPP_INCLUDED
+#define LBANN_EXECUTION_ALGORITHMS_LTFB_MUTATION_STRATEGY_HPP_INCLUDED
+
+#include "lbann/utils/random.hpp"
+#include "lbann/models/model.hpp"
+#include "lbann/layers/math/unary.hpp"
+#include "lbann/layers/activations/relu.hpp"
+#include "lbann/layers/activations/softmax.hpp"
+#include "lbann/layers/activations/elu.hpp"
+#include "lbann/layers/activations/leaky_relu.hpp"
+#include "lbann/layers/activations/log_softmax.hpp"
+
+namespace lbann {
+namespace ltfb {
+
+class MutationStrategy : public Cloneable<HasAbstractFunction<MutationStrategy>>
+{
+public:
+  MutationStrategy() {};
+  virtual ~MutationStrategy() = default;
+
+public:
+  /** @brief Apply a change to the model.
+   *  @param[in,out] m The model to change.
+   *  @param[in] step The current execution step in LTFB
+   */
+  virtual void mutate(model& m, const int& step) = 0;
+};
+
+// No Mutation
+class NullMutation : public Cloneable<NullMutation, MutationStrategy>
+{
+
+public:
+  NullMutation() = default; 
+  void mutate(model&, const int&) override {}
+};
+
+// Replace activation layers
+class ReplaceActivation : public Cloneable<ReplaceActivation, MutationStrategy>
+{
+private:
+  std::vector<std::string> m_activation_types = {"relu", "tanh", "softmax", "elu", "leaky relu", "log softmax"};
+
+public:
+  ReplaceActivation() = default;
+
+  std::unique_ptr<lbann::Layer> make_new_activation_layer(lbann::lbann_comm& comm,
+                                                          std::string const& new_type,
+                                                          std::string const& new_name)
+  {
+    using DataType = float;
+    #ifdef LBANN_HAS_GPU
+      constexpr El::Device Dev = El::Device::GPU;
+    #else
+      constexpr El::Device Dev = El::Device::CPU;
+    #endif
+
+    std::unique_ptr<lbann::Layer> layer;
+
+    if(new_type == "relu") {
+       layer = std::make_unique<
+           lbann::relu_layer<DataType, data_layout::DATA_PARALLEL, Dev>>(&comm); 
+    } else if (new_type == "tanh") {
+       layer = std::make_unique<
+            lbann::tanh_layer<DataType, data_layout::DATA_PARALLEL, Dev>>(&comm);
+    } else if (new_type == "softmax") {
+       layer = std::make_unique<
+            lbann::softmax_layer<DataType, data_layout::DATA_PARALLEL, Dev>>(&comm, softmax_mode::INSTANCE);
+    } else if (new_type == "elu") {
+       layer = std::make_unique<
+            lbann::elu_layer<DataType, data_layout::DATA_PARALLEL, Dev>>(&comm, 1);
+    } else if (new_type == "leaky relu") {
+       layer = std::make_unique<
+            lbann::leaky_relu_layer<DataType, data_layout::DATA_PARALLEL, Dev>>(&comm, 0.01);
+    } else if (new_type == "log softmax") {
+       layer = std::make_unique<
+            lbann::log_softmax_layer<DataType, data_layout::DATA_PARALLEL, Dev>>(&comm);
+    } else {
+       LBANN_ERROR("Unknown new layer type: ", new_type);
+    }
```

Comment:
Shouldn't this inherit the type, layout, and device allocation of the layer it's replacing?

Reply:
Has to be resolved at compile time, so cant use getter functions.